eth/watchers/eventdecoder.go

```diff
@@ -0,0 +1,73 @@
+package watchers
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// EventDecoder decodes logs into events for known contracts
+type EventDecoder struct {
+	addr             ethcommon.Address
+	contract         *bind.BoundContract
+	topicToEventName map[ethcommon.Hash]string
+}
+
+// FindEventName returns the event name for a log. An error will be returned if the log is not emitted
+// from a known contract or if it does not map to a known event
+func (e *EventDecoder) FindEventName(log types.Log) (string, error) {
+	switch log.Address {
+	case e.addr:
+		return e.findEventName(log)
+	default:
+		return "", errors.New("log not from known contract")
+	}
+}
+
+// Decode decodes a log into an event struct. An error will be returned if the log is not emitted
+// from a known contract or if it does not map to a known event
+func (e *EventDecoder) Decode(eventName string, log types.Log, decodedLog interface{}) error {
+	switch log.Address {
+	case e.addr:
+		return e.decode(eventName, log, decodedLog)
+	default:
+		return errors.New("log not from known contract")
+	}
+}
+
+func NewEventDecoder(addr ethcommon.Address, abiJSON string) (*EventDecoder, error) {
```

Comment:
Small style nit: could we move this constructor up after the type definition so that the layout is:

```
type definition -> constructor -> methods
```

Reply:
fixed in ca5b5af